README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# USgas <a href='https://ramikrispin.github.io/USgas/'><img src='man/figures/USgas.png' align="right" width="150" height="150" /></a>

<!-- badges: start -->

[![CRAN_Status_Badge](https://www.r-pkg.org/badges/version/USgas)](https://cran.r-project.org/package=USgas)
[![lifecycle](https://img.shields.io/badge/lifecycle-experimental-orange.svg)](https://lifecycle.r-lib.org/articles/stages.html)
[![License:
MIT](https://img.shields.io/badge/License-MIT-blue.svg)](https://opensource.org/license/mit/)
[![GitHub
commit](https://img.shields.io/github/last-commit/RamiKrispin/USgas)](https://github.com/RamiKrispin/USgas/commit/main)

<!-- badges: end -->

The **USgas** package provides an overview of demand for natural gas in
the US in a time-series format. That includes the following dataset:

- `usgas` - The monthly consumption of natural gas in the US/state level
  by end-use since 1973 for US level and 1989 for state level. It
  includes the following end-use categories:

  - Commercial Consumption
  - Delivered to Consumers
  - Electric Power Consumption
  - Industrial Consumption
  - Lease and Plant Fuel Consumption
  - Pipeline Fuel Consumption
  - Residential Consumption
  - Vehicle Fuel Consumption

The package also includes the following datasets, from previous release:

- `us_total` - The US annual natural gas consumption by state-level
  between 1997 and 2019, and aggregate level between 1949 and 2019
- `us_monthly` - The monthly demand for natural gas in the US between
  2001 and 2020
- `us_residential` - The US monthly natural gas residential consumption
  by state and aggregate level between 1989 and 2020

The `us_total`, `us_monthly`, and `us_residential` can be derived out of
the `usgas` dataset. Therefore, those datasets in the process of
deprication and will be removed in the next release to CRAN.

Data source: The US Energy Information Administration
[API](https://www.eia.gov/)

More information about the package datasets available on this
[vignette](https://ramikrispin.github.io/USgas/articles/introduction.html).

## Installation

You can install the released version of USgas from
[CRAN](https://CRAN.R-project.org) with:

``` r
install.packages("USgas")
```

And the development version from [GitHub](https://github.com/) with:

``` r
# install.packages("devtools")
devtools::install_github("RamiKrispin/USgas")
```

## Example

Let’s load the data:

``` r
data("usgas")

head(usgas)
#>         date                 process state state_abb      y
#> 1 1973-01-01  Commercial Consumption  U.S.      U.S. 392315
#> 2 1973-01-01 Residential Consumption  U.S.      U.S. 843900
#> 3 1973-02-01  Commercial Consumption  U.S.      U.S. 394281
#> 4 1973-02-01 Residential Consumption  U.S.      U.S. 747331
#> 5 1973-03-01  Commercial Consumption  U.S.      U.S. 310799
#> 6 1973-03-01 Residential Consumption  U.S.      U.S. 648504

str(usgas)
#> 'data.frame':    92783 obs. of  5 variables:
#>  $ date     : Date, format: "1973-01-01" "1973-01-01" ...
#>  $ process  : chr  "Commercial Consumption" "Residential Consumption" "Commercial Consumption" "Residential Consumption" ...
#>  $ state    : chr  "U.S." "U.S." "U.S." "U.S." ...
#>  $ state_abb: chr  "U.S." "U.S." "U.S." "U.S." ...
#>  $ y        : int  392315 843900 394281 747331 310799 648504 231943 465867 174258 326313 ...
#>  - attr(*, "units")= chr "MMCF"
#>  - attr(*, "product_name")= chr "Natural Gas"
#>  - attr(*, "source")= chr "EIA API: https://www.eia.gov/opendata/browser/natural-gas"
```

Plotting the residential consumption of natural gas in the US:

``` r
library(plotly)

us_res <- usgas[which(usgas$state == "U.S." & usgas$process == "Residential Consumption"), ]

plot_ly(data = us_res,
        x = ~ date,
        y = ~ y,
        type = "scatter",
        mode = "line") |>
  layout(title = "US Monthly Residential Consumption",
         yaxis = list(title = "MMCF"),
         xaxis = list(title = "Source: EIA API"))
```

<img src="man/figures/us_res.svg" width="100%" />

Plotting the total monthly natural gas delivered in the New England
states:

``` r
ne <- c("Connecticut", "Maine", "Massachusetts",
        "New Hampshire", "Rhode Island", "Vermont")
ne_gas <-  usgas[which(usgas$state %in% ne & usgas$process == "Delivered to Consumers"),]

plot_ly(data = ne_gas,
        x = ~ date,
        y = ~ y,
        color = ~ state,
        type = "scatter",
        mode = "line") |>
  layout(title = "Total Natrual Gas Delivered to Consumers in New England States",
         yaxis = list(title = "MMCF"),
         xaxis = list(title = "Source: EIA API"))
```

<img src="man/figures/new_england.svg" width="100%" />


NEWS.md
# USgas 0.1.2

* Add new dataset `usgas`, describing the US monthly natural gas consumption by end-use on the state and US aggregate level
* Deprecating the previous package datasets - `us_residential`, `us_total`, and `us_total`, as they can be generated by the `usgas` dataset

# USgas 0.1.1

* Fixing CRAN errors - removing the `digits` argument from the `format` function
* Updating the data - monthly data up-to-date to Feb 21

# USgas 0.1.0

* Added the `us_residential` dataset, describing the US monthly natural gas residential consumption by state and aggregate level
* Added the `us_total` dataset, describing the US annual natural gas consumption by state and aggregate level
* Added the `us_monthly` dataset, describing the overall US monthly natural gas consumption
* Added a `NEWS.md` file to track changes to the package.


R/usgas.R
#' US Monthly Consumption of Natural Gas by End Use
#' @description The US monthly consumption of natural gas by end-use and state between 1973 (US aggregate, state level since 1989) and 2022.
#'
#' Units: Million Cubic Feet
#'
#'
#' @format A data.frame with 6 variables.
#' \describe{
#'   \item{date}{A Date, the month and year of the observation (the day set by default to 1st of the month)}
#'   \item{process}{The process type description}
#'   \item{state}{The US state name}
#'   \item{state_abb}{the US state abbreviation}
#'   \item{y}{A numeric, the monthly natural gas residential consumption in a million cubic feet}
#'   }
#' @source US Energy Information Administration (EIA) \href{https://www.eia.gov/}{website}.
#' @keywords datasets timeseries natural gas us state
#' @details The dataset contains monthly summary of the consumption of natural gas
#' by end-use in the US by state and total aggregate level. The data is available
#' for the state level between January 1989 and December 2022, and for the US level
#' between January 1973 and Dec 2022. It includes the following end-use categories:
#'
#' - Commercial Consumption
#' - Delivered to Consumers
#' - Electric Power Consumption
#' - Industrial Consumption
#' - Lease and Plant Fuel Consumption
#' - Pipeline Fuel Consumption
#' - Residential Consumption
#' - Vehicle Fuel Consumption
#'
#' @examples
#'
#' library(plotly)
#'
#' data("usgas")
#'
#' head(usgas)
#'
#' # Plot the US consumption
#'
#' us_df <- usgas[which(usgas$state == "U.S."), ]
#'
#' plot_ly(data = us_df,
#'         x = ~ date,
#'         y = ~ y,
#'         color = ~ process,
#'         type = "scatter",
#'         mode = "line") |>
#'   layout(title = "US Monthly Consumption by End Use",
#'          yaxis = list(title = "MMCF"),
#'          xaxis = list(title = "Source: EIA Website"),
#'          legend = list(x = 0, y = 1.05),
#'          margin = list(l = 50, r = 50, b = 70, t = 60))
#'
#'
#' # Plot the California consumption
#'
#' ca_df <- usgas[which(usgas$state == "California"), ]
#'
#' plot_ly(data = ca_df,
#'         x = ~ date,
#'         y = ~ y,
#'         color = ~ process,
#'         type = "scatter",
#'         mode = "line") |>
#'   layout(title = "California Monthly Consumption by End Use",
#'          yaxis = list(title = "MMCF"),
#'          xaxis = list(title = "Source: EIA Website"),
#'          legend = list(x = 0, y = 1.05),
#'          margin = list(l = 50, r = 50, b = 70, t = 60))
#'
"usgas"


data_raw/data_raw.R
# Pulling US Monthly Consumption of Natural Gas

# Source: https://www.eia.gov/opendata/browser/natural-gas/cons/sum?frequency=monthly&data=value;&facets=product;process;&product=EPG0;&process=VRS;&end=2022-12&sortColumn=period;&sortDirection=desc;

eia_key <- Sys.getenv("eia_key")

years <- 1973:2022

url <- "https://api.eia.gov/v2/natural-gas/cons/sum/data/"


# US consumption by sector type

usgas_raw <- lapply(years, function(i){
  print(i)
  gas_meta <- EIAapi::eia_get(api_key = eia_key ,
                              api_url = url,
                              format = "data.frame",
                              facets = list(process = c("VCS",
                                                        "VDV",
                                                        "VEU",
                                                        "VGL",
                                                        "VGP",
                                                        "VGT",
                                                        "VIN",
                                                        "VRS"),
                                            product = "EPG0"),
                              start = paste(i, "-01", sep = ""),
                              end = paste(i + 1, "-01", sep = ""),
                              frequency = "monthly",
                              length = 5000,
                              offset = 0)

  return(gas_meta)

}) |> dplyr::bind_rows()

unique(usgas_raw$units)

usgas_meta <- usgas_raw |>
  dplyr::select(state_raw = `area-name`) |>
  dplyr::distinct() |>
  dplyr::mutate(state_abb = gsub(pattern = "USA-", replacement = "", x = state_raw),
                state_abb = ifelse(state_abb == "CALIFORNIA", "CA", state_abb),
                state_abb = ifelse(state_abb == "COLORADO", "CO", state_abb),
                state_abb = ifelse(state_abb == "OHIO", "OH", state_abb),
                state_abb = ifelse(state_abb == "NEW YORK", "NY", state_abb),
                state_abb = ifelse(state_abb == "FLORIDA", "FL", state_abb),
                state_abb = ifelse(state_abb == "MASSACHUSETTS", "MA", state_abb),
                state_abb = ifelse(state_abb == "WASHINGTON", "WA", state_abb),
                state_abb = ifelse(state_abb == "MINNESOTA", "MN", state_abb),
                state_abb = ifelse(state_abb == "TEXAS", "TX", state_abb),
                state_abb = ifelse(state_abb == "TEXAS", "TX", state_abb),
                state_name = state.name[match(state_abb,state.abb)],
                state_name = ifelse(state_abb == "DC", "District of Columbia", state_name),
                state_name = ifelse(state_abb == "U.S.", "U.S.", state_name))



usgas <- usgas_raw |>
  dplyr::mutate(date = as.Date(paste(period, "-01", sep = ""))) |>
  dplyr::select(date, state_raw = `area-name`,
                # process,
                process = `process-name`,
                y = value) |>
  dplyr::left_join(usgas_meta |>
                     dplyr::select(state_raw, state = state_name, state_abb),
                   by = "state_raw") |>
  dplyr::select(-state_raw) |>
  dplyr::select(date, process, state, state_abb, y) |>
  dplyr::arrange(date, process)

head(usgas)

attr(usgas, "units") <- "MMCF"
attr(usgas, "product_name") <- "Natural Gas"
attr(usgas, "source") <- "EIA API: https://www.eia.gov/opendata/browser/natural-gas"

max(usgas$date)

usethis::use_data(usgas, overwrite = TRUE)
write.csv(usgas, "./csv/usgas.csv", row.names = FALSE)


R/data.R
#' US Monthly Natural Gas Residential Consumption
#' @description The US monthly consumption of natural gas by residential consumers by state and aggregate level between 1989 and 2020
#'
#' Units: Million Cubic Feet
#'
#'
#' @format A data.frame with 3 variables.
#' \describe{
#'   \item{date}{A Date, the month and year of the observation (the day set by default to 1st of the month)}
#'   \item{state}{A character, the US state indicator}
#'   \item{y}{A numeric, the monthly natural gas residential consumption in a million cubic feet}
#'   }
#' @source US Energy Information Administration (EIA) \href{https://www.eia.gov/}{website}.
#' @keywords datasets timeseries natural gas us state
#' @details The dataset contains monthly summary of the demand for natural gas by residential consumers in the US by state and total aggregate level. The data is available for the state level between January 1989 and September 2020, and for the US level between January 1973 and September 2020.
#'
#' @examples
#' data("us_residential")
#'
#' # Subsetting the total consumption in the US
#'
#' us_res_total <- us_residential[which(us_residential$state == "U.S."),]
#'
#' us_res_total <- us_res_total[order(us_res_total$date), ]
#'
#' head(us_res_total)
#'
#' at_x <- seq.Date(from = as.Date("1980-01-01"),
#'                  to = as.Date("2020-01-01"),
#'                  length.out = 5)
#'
#' at_y <- pretty(us_res_total$y)[c(2, 4, 6)]
#'
#' plot(us_res_total$date, us_res_total$y,
#'      col = "#1f77b4",
#'      type = "l",
#'      frame.plot = FALSE,
#'      axes = FALSE,
#'      panel.first = abline(h = at_y, col = "grey80"),
#'      main = "US Natural Gas Residential Consumption",
#'      xlab = "Source: https://www.eia.gov/",
#'      ylab = "Million Cubic Feet")
#'
#'
#'
#' mtext(side =1, text = format(at_x, format = "%Y %b"), at = at_x,
#'       col = "grey20", line = 1, cex = 0.8)
#'
#' mtext(side =2, text = format(at_y, scientific = FALSE), at = at_y,
#'       col = "grey20", line = 1, cex = 0.8)
#'
#' ### Plotting the west cost consumption
#' # Subsetting the west cost states
#' wc_gas <- us_residential[which(us_residential$state %in%
#'                              c("Alaska", "California",
#'                                "Oregon", "Washington")),]
#'
#' # Reshape to wide
#' wc_wide <- reshape(wc_gas, v.names = "y", idvar = "date",
#'                    timevar = "state", direction = "wide")
#'
#' names(wc_wide) <- c("date","Alaska", "California",
#'                     "Oregon", "Washington")
#' # Reorder the data
#' wc_wide <- wc_wide[order(wc_wide$date), ]
#'
#' # Set the plot y and x axis ticks
#' at_x <- seq.Date(from = as.Date("1990-01-01"),
#'                  to = as.Date("2020-01-01"),
#'                  length.out = 4)
#'
#' at_y <- pretty(wc_gas$y)[c(2, 4, 6)]
#'
#' # plot the first series
#' plot(wc_wide$date, wc_wide$Alaska,
#'      type = "l",
#'      frame.plot = FALSE,
#'      axes = FALSE,
#'      panel.first = abline(h = c(at_y), col = "grey80"),
#'      main = "West Cost Natural Gas Residential Consumption",
#'      cex.main = 1, font.main = 1, col.main = "black",
#'      xlab = "Source: https://www.eia.gov/",
#'      font.axis = 1, cex.lab= 0.7,
#'      ylab = "Million Cubic Feet",
#'      ylim = c(min(wc_gas$y, na.rm = TRUE), max(wc_gas$y, na.rm = TRUE)))
#'
#' # Add the 3 other series
#' lines(wc_wide$date, wc_wide$California, col = "#1f77b4")
#' lines(wc_wide$date, wc_wide$Oregon, col = "#457b9d")
#' lines(wc_wide$date, wc_wide$Washington, col = "#ef476f")
#'
#' # Add the y and x axis ticks
#' mtext(side =1, text = format(at_x, format = "%Y %b"), at = at_x,
#'       col = "grey20", line = 1, cex = 0.8)
#'
#' mtext(side =2, text = format(at_y, scientific = FALSE), at = at_y,
#'       col = "grey20", line = 1, cex = 0.8)
#'
#' text(tail(wc_wide$date, 10)[1],
#'      max(wc_gas$y, na.rm = TRUE),
#'      "California",
#'      col = "#1f77b4",
#'      cex = 0.6)
#'
#' text(tail(wc_wide$date, 10)[1],
#'      max(wc_gas$y, na.rm = TRUE) - 5000,
#'      "Washington",
#'      col = "#ef476f",
#'      cex = 0.6)
#'
#' text(tail(wc_wide$date, 10)[1],
#'      max(wc_gas$y, na.rm = TRUE) - 10000,
#'      "Oregon",
#'      col = "#457b9d",
#'      cex = 0.6)
#'
#' text(tail(wc_wide$date, 10)[1],
#'      max(wc_gas$y, na.rm = TRUE) - 15000,
#'      "Alaska",
#'      col = "black",
#'      cex = 0.6)


"us_residential"

#' US Annual Total Natural Gas Consumption
#' @description The dataset provides information on the US annual natural gas consumption by state level between 1997 and 2019, and aggregate level between 1949 and 2019
#'
#' Units: Million Cubic Feet
#'
#'
#' @format A data.frame with 3 variables.
#' \describe{
#'   \item{year}{An integer, the year of the observation}
#'   \item{state}{A character, the US state indicator}
#'   \item{y}{A numeric, the yearly total natural gas consumption in a million cubic feet by state or US aggregate}
#'   }
#' @source US Energy Information Administration (EIA) \href{https://www.eia.gov/}{website}.
#' @keywords datasets timeseries natural gas us state
#' @details The dataset contains a summary of the total annual demand for natural gas in the US by state and total aggregate level. The data is available for the state level between 1997 and 2019, and for the US level between 1949 and 2019.
#'
#' @examples
#' data("us_total")
#'
#' # Subsetting the total consumption in the US
#' us_agg <- us_total[which(us_total$state == "U.S."),]
#'
#' us_agg <- us_agg[order(us_agg$year), ]
#'
#' head(us_agg)
#'
#' at_x <- seq(from = 1950,
#'             to = 2020,
#'             by = 10)
#'
#' at_y <- pretty(us_agg$y)[c(3, 5, 7)]
#'
#' plot(us_agg$year, us_agg$y,
#'      col = "#1f77b4",
#'      type = "l",
#'      frame.plot = FALSE,
#'      axes = FALSE,
#'      panel.first = abline(h = at_y, col = "grey80"),
#'      main = "US Total Natural Gas Consumption",
#'      xlab = "Source: https://www.eia.gov/",
#'      ylab = "Million Cubic Feet")
#'
#'
#'
#' mtext(side =1, text = format(at_x,  nsmall=0), at = at_x,
#'       col = "grey20", line = 1, cex = 0.8)
#'
#' mtext(side =2, text = format(at_y, scientific = FALSE), at = at_y,
#'       col = "grey20", line = 1, cex = 0.8)
#'
#' ### Plotting the annual consumption in New England states
#' # Subsetting the New England states
#' ne <- c("Connecticut", "Maine", "Massachusetts",
#'         "New Hampshire", "Rhode Island", "Vermont")
#' ne_gas <-  us_total[which(us_total$state %in% ne),]
#'
#' # Reshape to wide
#' ne_wide <- reshape(ne_gas, v.names = "y", idvar = "year",
#'                    timevar = "state", direction = "wide")
#'
#' names(ne_wide) <- c("year",ne)
#' # Reorder the data
#' ne_wide <- ne_wide[order(ne_wide$year), ]
#'
#' # Set the plot y and x axis ticks
#' at_x <- seq(from = 2000,
#'             to = 2020,
#'             by = 5)
#'
#' at_y <- pretty(ne_gas$y)[c(2, 4, 6)]
#'
#' # plot the first series
#' plot(ne_wide$year, ne_wide$Connecticut,
#'      type = "l",
#'      col = "#073b4c",
#'      frame.plot = FALSE,
#'      axes = FALSE,
#'      panel.first = abline(h = c(at_y), col = "grey80"),
#'      main = "New England Annual Natural Gas Consumption by State",
#'      cex.main = 1, font.main = 1, col.main = "black",
#'      xlab = "Source: https://www.eia.gov/",
#'      font.axis = 1, cex.lab= 0.7,
#'      ylab = "Million Cubic Feet",
#'      ylim = c(min(ne_gas$y, na.rm = TRUE), max(ne_gas$y, na.rm = TRUE)),
#'      xlim = c(min(ne_gas$year), max(ne_gas$year) + 3))
#'
#' # Add the 5 other series
#' lines(ne_wide$year, ne_wide$Maine, col = "#1f77b4")
#' lines(ne_wide$year, ne_wide$Massachusetts, col = "#118ab2")
#' lines(ne_wide$year, ne_wide$`New Hampshire`, col = "#06d6a0")
#' lines(ne_wide$year, ne_wide$`Rhode Island`, col = "#ffd166")
#' lines(ne_wide$year, ne_wide$Vermont, col = "#ef476f")
#'
#' # Add the y and x axis ticks
#' mtext(side =1, text = format(at_x, nsmall=0), at = at_x,
#'       col = "grey20", line = 1, cex = 0.8)
#'
#' mtext(side =2, text = format(at_y, scientific = FALSE), at = at_y,
#'       col = "grey20", line = 1, cex = 0.8)
#'
#'
#' text(max(ne_wide$year) + 2,
#'      tail(ne_wide$Connecticut,1),
#'      "Connecticut",
#'      col = "#073b4c",
#'      cex = 0.7)
#'
#' text(max(ne_wide$year) + 2,
#'      tail(ne_wide$Maine,1) * 0.95,
#'      "Maine",
#'      col = "#1f77b4",
#'      cex = 0.7)
#'
#' text(max(ne_wide$year) + 2,
#'      tail(ne_wide$Massachusetts,1),
#'      "Massachusetts",
#'      col = "#118ab2",
#'      cex = 0.7)
#'
#' text(max(ne_wide$year) + 2,
#'      tail(ne_wide$`New Hampshire`,1) * 1.1,
#'      "New Hampshire",
#'      col = "#06d6a0",
#'      cex = 0.7)
#'
#' text(max(ne_wide$year) + 2,
#'      tail(ne_wide$`Rhode Island`,1) * 1.05,
#'      "Rhode Island",
#'      col = "#ffd166",
#'      cex = 0.7)
#'
#' text(max(ne_wide$year) + 2,
#'      tail(ne_wide$Vermont,1),
#'      "Vermont",
#'      col = "#ef476f",
#'      cex = 0.7)

"us_total"

#' US Monthly Total Natural Gas Consumption
#' @description The dataset provides information about the total monthly natural gas consumption in the US between 2001 and 2020
#'
#' Units: Million Cubic Feet
#'
#'
#' @format A data.frame with 2 variables.
#' \describe{
#'   \item{date}{A Date, the month and year of the observation (the day set by default to 1st of the month)}
#'   \item{y}{A numeric, the US monthly total natural gas consumption in a million cubic feet}
#'   }
#' @source US Energy Information Administration (EIA) \href{https://www.eia.gov/}{website}.
#' @keywords datasets timeseries natural gas us state
#' @details The dataset contains monthly summary of the total demand for natural gas in the US between 2001 and 2020.
#'
#' @examples
#'
#' data("us_monthly")
#'
#' head(us_monthly)
#'
#' at_x <- seq.Date(from = as.Date("2000-01-01"),
#'                  to = as.Date("2020-01-01"),
#'                  by = "5 years")
#'
#' at_y <- pretty(us_monthly$y)[c(2, 4, 6)]
#'
#' plot(us_monthly$date, us_monthly$y,
#'      col = "#1f77b4",
#'      type = "l",
#'      frame.plot = FALSE,
#'      axes = FALSE,
#'      panel.first = abline(h = at_y, col = "grey80"),
#'      main = "US Total Natural Gas Consumption",
#'      xlab = "Source: https://www.eia.gov/",
#'      ylab = "Million Cubic Feet")
#'
#'
#'
#' mtext(side =1, text = format(at_x, format = "%Y"), at = at_x,
#'       col = "grey20", line = 1, cex = 0.8)
#'
#' mtext(side =2, text = format(at_y, scientific = FALSE), at = at_y,
#'       col = "grey20", line = 1, cex = 0.8)
#'

"us_monthly"
